TS-4399 TS-4400 Management API messes up proxy options

danobi opened the pull request:
TS-4399: Management API breaks diagnostic log rotation
TS-4400: TSProxyStateSet persist cache clearing across restart

The two issues are related in that they both deal with the
management API not correctly handling proxy flags.

For TS-4399, it was because the management API was not aware
of traffic_manager setting extra proxy options. This was fixed
by providing CoreAPI a callback to get extra proxy options from
traffic_manager.

For TS-4400, it was because the management API was not properly
clearing optional flags between proxy reboots. This was fixed
by resetting the proxy options before each reboot.

danobi commented:
@jpeach, is this what you were thinking?

Also, between commit 82037 and 3365b, `traffic_ctl` stopped working immediately after ATS starts up. There is ~30s gap between startup any `traffic_ctl` commands working, eg. `sudo ./traffic_ctl server stop`. I'm still looking into this.

danobi commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -835,10 +824,11 @@ main(int argc, const char **argv)
         just_started = 0;
         sleep_time   = 0;
       } else {
+        mgmt_log("in ProxyStateSet else branch");
```
Note to self: remove this line

jpeach replied:
Yes please :)

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -610,32 +616,37 @@ main(int argc, const char **argv)
-    snprintf(lmgmt->proxy_options + l, n, " --%s %s", TM_OPT_BIND_STDERR, bind_stderr);
+    char *bind_stderr_opt = new char[strlen("--") + strlen(TM_OPT_BIND_STDERR)];
+    strcpy(bind_stderr_opt, "--");
+    strcat(bind_stderr_opt, TM_OPT_BIND_STDERR);
```
```C
foo.push_back("--" TM_OPT_BIND_STDERR")

danobi commented:
How does that look? Still need to figure out that delay issue.

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -27,6 +27,7 @@
 #include "ts/ink_sock.h"
 #include "ts/ink_args.h"
 #include "ts/ink_syslog.h"
+#include <vector>
```
Remove this.

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -445,6 +446,8 @@ main(int argc, const char **argv)
   int binding_version      = 0;
   BindingInstance *binding = NULL;
 
+  std::vector<char*> callback_proxy_options;
```
Remove this.

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -611,31 +614,35 @@ main(int argc, const char **argv)
+    strcpy(new_proxy_opts, lmgmt->proxy_options);
+    strcat(new_proxy_opts, bind_stdout_opt);
+    strcat(new_proxy_opts, bind_stdout);
+    delete lmgmt->proxy_options;
```
You can replace all this string code with ``TextBuffer``.

```C
TextBuffer args;

if (*bind_stdout) {
   const char * space = args.empty() ? "" : " ";
   args.format("%s%s", space, "--" TM_OPT_BIND_STDOUT);
}

...

lmgmt->proxy_options = args.release();
```

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -611,31 +614,35 @@ main(int argc, const char **argv)
     ats_free(lmgmt->proxy_options);
-    lmgmt->proxy_options = tsArgs;
-    mgmt_log("[main] Traffic Server Args: '%s'\n", lmgmt->proxy_options);
+    lmgmt->proxy_options = new_proxy_opts;
```
At this point ``lmgmt->proxy_options`` isn't set, so just copy ``tsArgs`` into the text buffer if it is present.

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -611,31 +614,35 @@ main(int argc, const char **argv)
+    strcpy(new_proxy_opts, lmgmt->proxy_options);
+    strcat(new_proxy_opts, bind_stderr_opt);
+    strcat(new_proxy_opts, bind_stderr);
+    delete lmgmt->proxy_options;
```
You should not mix new/delete and malloc/free here. But this problem goes away with the text buffer.

jpeach commented on cmd/traffic_manager/traffic_manager.cc:
```diff
@@ -844,6 +852,11 @@ main(int argc, const char **argv)
+  // This probably isn't needed b/c the process dies here...
+  for (auto it = callback_proxy_options.begin(); it < callback_proxy_options.end(); ++it)
+    free(*it);
+
```
Remove this.

jpeach commented on mgmt/LocalManager.cc:
```diff
@@ -239,8 +239,9 @@ LocalManager::LocalManager(bool proxy_on) : BaseManager(), run_proxy(proxy_on),
-  proxy_options                = REC_readString("proxy.config.proxy_binary_opts", &found);
   env_prep                     = REC_readString("proxy.config.env_prep", &found);
+  proxy_options                = new char[1];
+  strcpy(proxy_options, "");  // so later we can always `delete proxy_options` without worrying
```
You don't need this. It is safe for ``proxy_options`` to be NULL.

jpeach commented on mgmt/api/CoreAPI.cc:
```diff
@@ -41,6 +41,7 @@
 #include "ts/Diags.h"
 #include "ts/ink_hash_table.h"
 #include "ExpandingArray.h"
+#include <vector>
```
Remove this.

jpeach commented on mgmt/api/CoreAPI.cc:
```diff
@@ -53,6 +54,9 @@
 CallbackTable *local_event_callbacks;
 
+// Used to get any proxy options that traffic_manager might want to tack on
+std::function<void(std::vector<char*>&)> proxy_options_callback;
```
Remove this.

jpeach commented on mgmt/LocalManager.h:
```diff
@@ -74,7 +74,7 @@ class LocalManager : public BaseManager
 
   void processEventQueue();
-  bool startProxy();
+  bool startProxy(char *onetime_options);
```
``const char *``.

jpeach commented on mgmt/LocalManager.cc:
```diff
@@ -831,9 +832,13 @@ LocalManager::processEventQueue()
  */
 bool
-LocalManager::startProxy()
+LocalManager::startProxy(char *onetime_options)
```
``const char *``

jpeach commented on mgmt/LocalManager.cc:
```diff
@@ -902,8 +907,11 @@ LocalManager::startProxy()
 
     real_proxy_options.append(proxy_options, strlen(proxy_options));
+    real_proxy_options.append(" ", strlen(" "));
+    real_proxy_options.append(onetime_options, strlen(onetime_options));
```
Let's allow ``onetime_options`` to be NULL, so
```C
if (onetime_options && *onetime_options) {
  ...
}
```

danobi commented:
I think the delay issue is either:
1) The `mgmt_sleep_sec(1)` loop in `ProxyStateSet` b/c the main `traffic_manager` loop needs to run for the `lmgmt->proxy_running` flag to flip
2) Some kind of quirk with running `lmgmt->listenForProxy()`

danobi commented:
The actual issue was suspected issue #1, where there was a sleeping deadlock scenario. See commit message for details.

Other than that most recent change, I think the patch is ready for final review.

jpeach commented on mgmt/api/CoreAPI.cc:
```diff
@@ -184,21 +184,19 @@ ProxyStateSet(TSProxyStateT state, TSCacheClearT clear)
 
     lmgmt->run_proxy = true;
     lmgmt->listenForProxy();
+    lmgmt->startProxy(tsArgs);
```
Do you actually need to do the sleeping stuff here? Since you now call ``LocalManager:: startProxy()`` directly, you have the return value to know that it succeeded. I don't think that the contract for this API needs to include waiting for a message.

```C
return lmgmt->startProxy(tsArgs) ? TS_ERR_OKAY : TS_ERR_FAIL;
```

danobi replied:
That's a good observation, I'm kicking myself for now seeing it before.

jpeach commented on mgmt/LocalManager.cc:
```diff
@@ -902,8 +907,13 @@ LocalManager::startProxy()
 
-    if (!strstr(proxy_options, MGMT_OPT)) { // Make sure we're starting the proxy in mgmt mode
+    // Make sure we're starting the proxy in mgmt mode
+    if (!strstr(proxy_options, MGMT_OPT) && !strstr(onetime_options, MGMT_OPT)) {
```
Prefer ``strstr(...) == 0``. It's just that little bit more readable.

jpeach commented on mgmt/LocalManager.cc:
```diff
@@ -189,6 +189,7 @@ LocalManager::LocalManager(bool proxy_on) : BaseManager(), run_proxy(proxy_on),
   proxy_launch_outstanding  = false;
   mgmt_shutdown_outstanding = MGMT_PENDING_NONE;
   proxy_running             = 0;
+  coreapi_sleep             = true;
```
As mentioned in the other comment, I think we can get away without this.

jpeach commented:
[approve ci]

atsci commented:
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1157/ for details.

atsci commented:
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1050/ for details.

jpeach commented:
@danobi Since this covers 2 JIRAs, please designate one of them for the fix, and mark the other as a duplicate.

danobi commented:
I couldn't figure out where the dup option was so I changed TS-4400 to a subtask of TS-4399. 

Also, can you kick off the CI again? I think I fixed the build.

SolidWallOfCode commented on mgmt/LocalManager.cc:
```diff
@@ -902,8 +906,13 @@ LocalManager::startProxy()
 
-    if (!strstr(proxy_options, MGMT_OPT)) { // Make sure we're starting the proxy in mgmt mode
+    // Make sure we're starting the proxy in mgmt mode
+    if (strstr(proxy_options, MGMT_OPT) == 0 && strstr(onetime_options, MGMT_OPT) == 0) {
```
Couldn't you just check `real_proxy_options` here? If `onetime_options` aren't already in there, they won't be added anyway.

danobi replied:
`real_proxy_options` is the mysterious `Vec` class. AFAICT, I suppose we could, but it would require more allocation.
